dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,662 @@
+
+	// All match criteria must be satisfied before a set of devices will be used
+	// together.
+	//
+	// +optional
+	// +listType=atomic
+	Match []MatchModel `json:"match,omitempty"`
+
+	// DefaultRequests are individual requests for separate resources for a
+	// claim using this class. In contrast to configuration and
+	// requrirements, these requests are only used if the claim does not
+	// specify its own list of requests.
+	//
+	// If empty and the claim doesn't specify any requests either, then
+	// a single, empty request is used.
+	//
+	// +listType=atomic
+	DefaultRequests []ResourceRequest `json:"defaultRequests" protobuf:"bytes,5,name=requests"`
 }
 
-// DeviceClassDefinition specifies the subset of devices to consider part of
-// this class, along with device-specific configuration information for those
-// devices.
-type DeviceClassDefinition struct {
-	// Driver specifies the driver that should handle this class of devices.
-	// When a DeviceClaim uses this class, only devices published by the
-	// specified driver will be considered.
-	// +required
-	Driver string `json:driver,omitempty`
+// ConfigurationParameters must have one and only one field set.
+//
+// TBD: does this have to be a one-of? It might be useful to allow vendor configuration
+// in addition to in-tree configuration. Extending this type then becomes harder because
+// an old scheduler would silently ignore unknown fields.
+type ConfigurationParameters struct {
+	// +listType=atomic
+	Vendor []VendorConfigurationParameters `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
+}
 
-	// Constraints is a CEL expression that operates on device attributes,
-	// and must evaluate to true for a device to be considered. It will be
-	// ANDed with any Constraints field in the DeviceClaim using this class.
+// VendorConfigurationParameters contains configuration parameters for a driver.
+type VendorConfigurationParameters struct {
+	// DriverName is used to determine which kubelet plugin needs
+	// to be passed these configuration parameters.
+	//
+	// An admission webhook provided by the vendor could use this
+	// to decide whether it needs to validate them.
+	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
+
+	// Parameters can contain arbitrary data. It is the
+	// responsibility of the vendor to handle validation and
+	// versioning.
+	Parameters runtime.RawExtension `json:"parameters,omitempty" protobuf:"bytes,2,opt,name=parameters"`
+}
+
+// RequirementModel must have one and only one field set.
+type RequirementModel struct {
+	// Device describes a filter based on device attributes.
+	// This covers "qualititative" aspects of a device.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
+	Device *DeviceFilter `json:"device,omitempty"`
+
+	// Resource describes how much of a consumable resource provided by a
+	// device are needed.
+	//
+	// FUTURE EXTENSION, not planned for 1.31!
```

Comment:
I originally added this to show that a new entry for "consumable resources" could be added here with a non-breaking change because it goes into a one-of.

It would even work without a more complex resource model:
```
spec:
  requests:
  - requirements:
    - resource:
        name: memory.example.com-foozer
        quantity: 1Gi
```

This would be semantically equivalent to a CEL selector which checks for `memory.example.com-foozer >= 1Gi`. For devices with consumable resources the meaning can be extended to "consume this resource".

@klueska then went ahead and added the "shared resources" model that he and John were discussing. At that point, I am getting lost and have various questions about that model :sweat_smile: 

My suggestion is this:
- We don't discuss the shared resources model any further now.
- Instead, we use this PR to agree on the overall API.

Once we have that agreement, we merge it and use the directory to iterate further. In parallel, I will:
- Create a KEP update with the things that (hopefully) have consensus and that I can explain and know how to implement.
- We merge that KEP update and start implementing it.

Then while that goes on, @klueska and @johnbelamaric can hash out details of the shared resource model. If things go well, that can still go into 1.31 through another KEP update and a corresponding update of the implementation.

/cc @thockin

Reply:
I think we need to agree on the MVP *scope* and then the API. I'd really, really, like to close on the scope this week. At least, I would like the four of us - Patrick, Kevin, Tim, me - to be in alignment on the scope. We can then review it in the meeting on Tuesday next week. Next Monday is a holiday here.

Once we know the scope, we can stop arguing about things in the API that are not part of the 1.31 scope :)

Can we try to get that concluded? Doc is here: https://docs.google.com/document/d/1tvdBniYMWo79hrIeCYkJsUylzvh90xSOC9dm-rLAk88/edit

Reply:
Even if we agree that partionable devices are in scope for 1.31 (and I think we agree), I still don't want to discuss them in this PR. We need a staggered approach with the basics agreed upon first, then we can continue with the things "on the cusp".

Otherwise the risk is much too high that don't reach any agreement on next steps for 1.31.